src/meta.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * Managing `meta data` of each component that includes
+ * special method names and user defined decorator hooks. Since the property
+ * decorators are called before creating component constructor,
+ * we need to store decorator hooks in each prototype object and call it
+ * in component's class decorator later.
+ */
+
+import * as Vue from 'vue'
+import { Dictionary, VueInternal } from './declarations'
+
+export type DecoratorHook = (options: Vue.ComponentOptions<Vue>, value?: any) => void
+
+export class Meta {
+  /**
+   * The method names that the component decorator will
+   * just register on component options object.
+   * We can extend this list by using `Component.registerHooks`.
+   */
+  static internalHooks = [
+    'data', 'beforeCreate', 'created', 'beforeMount', 'mounted',
+    'beforeDestroy', 'destroyed', 'beforeUpdate', 'updated',
+    'activated', 'deactivated', 'render'
+  ]
+
+  /**
+   * Property, method and parameter decorators created by `createDecorator` helper
+   * will register functions that update component options for lazy processing.
+   * They will be executed just before creating component constructor.
+   */
+  decoratorMap: Dictionary<DecoratorHook> = {}
+
+  /**
+   * Property names that indicates the initial property will be
+   * processed on user-defined decorators. Also they will not collected in `data` hook.
+   */
+  decoratorAwaredKeys: Dictionary<boolean> = {}
+
+  get shouldGetInitalProperty () {
+    return Object.keys(this.decoratorAwaredKeys).length > 0
+  }
+}
+
+export interface CreateDecoratorOptions {
+  getInitialProperty?: boolean
+}
+
+// Property or method decorators
+export function createDecorator (
+  factory: (options: Vue.ComponentOptions<Vue>, key: string, value: any) => void,
+  options?: CreateDecoratorOptions
+): (target: Vue, key: string) => void
+
+// Parameter decorators
+export function createDecorator (
+  factory: (options: Vue.ComponentOptions<Vue>, key: string, index: number) => void,
+  options?: CreateDecoratorOptions
+): (target: Vue, key: string, index: number) => void
+
+export function createDecorator (
+  factory: (options: Vue.ComponentOptions<Vue>, key: string, index: number) => void,
+  options: CreateDecoratorOptions = {}
+): (target: VueInternal, key: string, index: any) => void {
+
+  return (proto, key, index) => {
+
+    let meta = proto.__vue_component_meta__
+    if (!meta) {
+      proto.__vue_component_meta__ = meta = new Meta()
+    }
+
+    // Allow to process an initial property on userland
+    if (options.getInitialProperty) {
+      meta.decoratorAwaredKeys[key] = true
+    }
+
+    meta.decoratorMap[key] = (options, value) => {
```

Comment:
Does this mean a property can only have one decorator applied?

If multiple decorators apply on the same decorator, only the last decorator's write to `decoratorMap` will be preserved, right?